core/src/main/java/com/linecorp/armeria/server/http/auth/OAuth1aToken.java

```diff
@@ -0,0 +1,216 @@
+     * oauth_signature_method parameter.
+     * */
+    private static final String OAUTH_SIGNATURE_METHOD = "oauth_signature_method";
+
+    /**
+     * oauth_signature parameter.
+     * */
+    private static final String OAUTH_SIGNATURE = "oauth_signature";
+
+    /**
+     * oauth_timestamp parameter.
+     * */
+    private static final String OAUTH_TIMESTAMP = "oauth_timestamp";
+
+    /**
+     * oauth_nonce parameter.
+     * */
+    private static final String OAUTH_NONCE = "oauth_nonce";
+
+    /**
+     * version parameter. (optional)
+     * If not set, the default value is 1.0.
+     * */
+    private static final String OAUTH_VERSION = "version";
+
+    /**
+     * Set of required parameters.
+     * */
+    private static final Set<String> REQUIRED_PARAM_KEYS = Sets.newHashSet(OAUTH_CONSUMER_KEY, OAUTH_TOKEN,
+                                                                           OAUTH_SIGNATURE_METHOD,
+                                                                           OAUTH_SIGNATURE, OAUTH_TIMESTAMP,
+                                                                           OAUTH_NONCE);
+
+    /**
+     * Set of optional parameters.
+     * */
+    private static final Set<String> OPTIONAL_PARAM_KEYS = Sets.newHashSet(REALM, OAUTH_VERSION);
+
+    /**
+     * Set of defined parameters, regardless of it is required or not.
+     * */
+    private static final Set<String> DEFINED_PARAM_KEYS = Sets.union(REQUIRED_PARAM_KEYS, OPTIONAL_PARAM_KEYS);
+
+    /**
+     * Creates a new {@link OAuth1aToken} from given arguments.
+     * */
+    public static OAuth1aToken of(Map<String, String> params) {
+        return new OAuth1aToken(params);
+    }
+
+    /**
+     * Map of parameters.
+     * */
+    private final Map<String, String> params;
+
+    private OAuth1aToken(Map<String, String> params) {
+        /**
+         * Map builder with default version value.
+         * */
+        ImmutableMap.Builder builder = ImmutableMap.<String, String>builder();
+
+        for (Entry<String, String> param : params.entrySet()) {
+            String key = param.getKey();
+            String value = param.getValue();
+
+            /**
+             * Empty values are ignored.
+             * */
+            if (!Strings.isNullOrEmpty(value)) {
+                String lowercased = key.toLowerCase(Locale.US);
+                if (DEFINED_PARAM_KEYS.contains(lowercased)) {
+                    /**
+                     * If given parameter is defined by Oauth1a protocol, add with lower-cased key.
+                     * */
+                    builder.put(lowercased, value);
+                } else {
+                    /**
+                     * Otherwise, just add.
+                     * */
+                    builder.put(key, value);
+                }
+            }
+        }
+
+        this.params = builder.build();
+
+        Set<String> missingParams = Sets.difference(REQUIRED_PARAM_KEYS, this.params.keySet());
```

Comment:
We don't need to create a new Set to check all required keys are present. (efficiency)